fem/tmop_pa_h2m.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "tmop.hpp"
+#include "tmop_pa.hpp"
+#include "linearform.hpp"
+#include "pgridfunc.hpp"
+#include "tmop_tools.hpp"
+#include "../general/forall.hpp"
+#include "../linalg/kernels.hpp"
+#include "../linalg/dtensor.hpp"
+
+namespace mfem
+{
+
+MFEM_REGISTER_TMOP_KERNELS(void, AddMultGradPA_Kernel_2D,
```

Comment:
``MFEM_REGISTER`` is new to me, what does it do?

Reply:
It registers the kernels into a static map that avoids all the `if` or `switch` tests.